src/taskrunner.py

```diff
@@ -0,0 +1,241 @@
+]
+
+DEBUG_TASK_LIST = [
+    aws.rds_snapshots,
+    aws.detailed_stack_list,
+
+    tasks.diff_builder_config,
+
+    deploy.load_balancer_status,
+    deploy.load_balancer_register_all,
+
+    master.write_missing_keypairs_to_s3,
+    master.download_keypair,
+    master.server_access,
+    master.remaster,
+    master.update_salt,
+    master.update_salt_master,
+    master.remaster_all,
+
+    buildvars.read,
+    buildvars.valid,
+    buildvars.fix,
+    buildvars.force,
+]
+
+def mk_task_map(task, qualified=True):
+    path = "%s.%s" % (task.__module__.split('.')[-1], task.__name__)
+    unqualified_path = task.__name__
+    description = (task.__doc__ or '').strip().replace('\n', ' ')[:60]
+    return {
+        "name": path if qualified else unqualified_path,
+        "path": path,
+        "description": description,
+        "fn": task,
+    }
+
+def generate_task_list(show_debug_tasks=False):
+    """returns a collated list of maps with task information.
+
+    [{"name": "fn", "fn": pathto.fn1, "description": "foo bar baz"}, ...]
+     {"name": "pathto.fn", "fn": pathto.fn2, "description": "foo bar baz"}, ...]"""
+
+    def to_list(task_list, qualified=True):
+        return [mk_task_map(task, qualified) for task in task_list]
+
+    new_task_list = to_list(UNQUALIFIED_TASK_LIST, qualified=False) + to_list(TASK_LIST)
+    if show_debug_tasks:
+        new_task_list = to_list(UNQUALIFIED_TASK_LIST, qualified=False) + \
+            to_list(UNQUALIFIED_DEBUG_TASK_LIST) + \
+            to_list(TASK_LIST) + \
+            to_list(DEBUG_TASK_LIST)
+
+    return new_task_list
+
+# --- taken from Fabric3 (fork of Fabric 1, BSD Licenced)
+# --- https://github.com/mathiasertl/fabric/blob/1.13.1/fabric/main.py#L499-L564
+def _escape_split(sep, argstr):
+    """
+    Allows for escaping of the separator: e.g. task:arg='foo\, bar'
+
+    It should be noted that the way bash et. al. do command line parsing, those
+    single quotes are required.
+    """
+    escaped_sep = r'\%s' % sep
+
+    if escaped_sep not in argstr:
+        return argstr.split(sep)
+
+    before, _, after = argstr.partition(escaped_sep)
+    startlist = before.split(sep)  # a regular split is fine here
+    unfinished = startlist[-1]
+    startlist = startlist[:-1]
+
+    # recurse because there may be more escaped separators
+    endlist = _escape_split(sep, after)
+
+    # finish building the escaped value. we use endlist[0] becaue the first
+    # part of the string sent in recursion is the rest of the escaped value.
+    unfinished += sep + endlist[0]
+
+    return startlist + [unfinished] + endlist[1:]  # put together all the parts
+
+def parse_arguments(arguments):
+    """
+    Parse string list into list of tuples: command, args, kwargs, hosts, roles.
+
+    See sites/docs/usage/fab.rst, section on "per-task arguments" for details.
+    """
+    cmds = []
+    for cmd in arguments:
+        args = []
+        kwargs = {}
+        if ':' in cmd:
```

Comment:
this is a tricky part that needs several unit tests. See https://github.com/elifesciences/elife-jenkins-workflow-libs/search?utf8=%E2%9C%93&q=bldr&type= for good examples of calls from Jenkins

Reply:
this was taken from fabric itself, it's what they were using to parse arguments.

When I realised bash was stripping the quotes and that there was no way to preserve them I replaced my implementation with theirs. The tests I had didn't change except to replicate the quote removal that bash was doing.

Commas need to be escaped, but this was the same under Fabric as well. It uses commas to separate commands.

I'll take a look at that link and see if any further tests need to be added

Reply:
The cases in the link should all be handled provided the values of the variables don't contain commas - which was true before as well, we just may not have realised it.